### ISSUE TITLE
Finetune qwen3-0.6b

### ISSUE THREAD

**Issue by bil-ash** [felladrin/gguf-trainer#41]
@felladrin Less of an issue and more of a question. I would like to finetune qwen3-0.6b on CPU on ryzen5pro 3600 with 32GB RAM with 10(out of the 12) threads dedicated to training. Is it feasible? If yes, what would be the peak memory usage?

**Comment by felladrin** [comment]
No, not with this repo. Training needs a WebGPU device (`pretrain` and `finetune` both exit with `no WebGPU: training needs Deno`), and the Ryzen 5 PRO 3600 has no GPU. The CPU backend here is a correctness reference, not a training path: it's single-threaded and there's no `--threads` flag, so the 10 threads wouldn't be used.

On memory, 32 GB wouldn't be enough anyway. Extrapolated from a measured 293M `qwen3` run (`pool 18053 + state 4495` at batch 2 x seq 2048):

| Shape | Peak |
| :--- | ---: |
| seq 512, batch 1 | ~20 GB |
| seq 1024, batch 1 | ~27 GB |
| seq 2048, batch 1 | ~33 GB |

About 15 GB of that is fixed before a single activation: f32 weights and gradients (4.8 GB) plus the Muon and AdamW state (9.9 GB).

For CPU, use [`transformers`](https://huggingface.co/docs/transformers) with [`peft`](https://huggingface.co/docs/peft). There's a page on [training on CPU](https://huggingface.co/docs/transformers/en/perf_train_cpu) too. I ran Qwen3-0.6B-Base this way to check, at seq 512 batch 1 with 10 threads: LoRA gives 118 tokens/s at 6.1 GB peak, a full fine-tune 82 tokens/s at 12.8 GB. Both fit your RAM. (I measured on a Zen 5 core with AVX-512, so on Zen 2 expect less.)

```sh
pip install torch --index-url https://download.pytorch.org/whl/cpu
pip install transformers peft
```

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import LoraConfig, get_peft_model

torch.set_num_threads(10)

tok = AutoTokenizer.from_pretrained("Qwen/Qwen3-0.6B-Base")
model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen3-0.6B-Base", dtype=torch.float32)
model.config.use_cache = False
model = get_peft_model(model, LoraConfig(
    r=16, lora_alpha=32, task_type="CAUSAL_LM",
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"]))

opt = torch.optim.AdamW([p for p in model.parameters() if p.requires_grad], lr=1e-4)
ids = tok("your training text here", return_tensors="pt").input_ids

model.train()
for step in range(100):
    loss = model(input_ids=ids, labels=ids).loss
    loss.backward()
    opt.step(); opt.zero_grad()
    print(step, loss.item())
```

Swap the loop for a [`Trainer`](https://huggingface.co/docs/transformers/en/main_classes/trainer) once it runs. And if you ever put a GPU in that machine, the 95M to 360M rows in the readme table are what this repo is actually good at.

**Comment by felladrin** [comment]
Coming back to correct one part of this: the memory numbers above are out of date.

The answer itself hasn't changed. Training here still needs a WebGPU device, so a Ryzen 5 PRO 3600 with no GPU still can't run it, and `transformers` with `peft` is still what I'd use on CPU.

But "32 GB wouldn't be enough anyway" is no longer true. Three flags landed since: `--loss-chunk` (fuses the readout into a chunked cross-entropy), `--recompute` (replays each layer in backward instead of storing its activations), and `--lora-rank`. Measured on the Qwen3-0.6B shape at seq-len 2048, batch 1:

| | peak |
| :--- | ---: |
| what I quoted above | ~33 GB |
| `--recompute --loss-chunk 8192` | **15.8 GB** |

So the memory argument doesn't hold. It's [lever 20](https://github.com/felladrin/gguf-trainer/blob/main/docs/optimization.md) if you want the numbers.

One other thing worth knowing, since it's the model you asked about: measuring this turned up a bug where any export above 2 GiB ran away and filled the disk, which meant Qwen3-0.6B couldn't be exported at all. Fixed in #50.

Sorry for the noise on a closed issue. The bottom line was right, but the table wasn't, and it's the kind of thing someone finds later and takes at face value.

**Comment by bil-ash** [comment]
Thanks
